src/main/kotlin/org/rust/ide/inspections/RustNamingInspection.kt

```diff
@@ -0,0 +1,253 @@
+        if (id == null) return
+        val (isOk, suggestedName) = checkName(id.text)
+        if (!isOk) {
+            holder.registerProblem(id, "$elementType `${id.text}` should have $styleName case name such as `$suggestedName`")
+        }
+    }
+
+    abstract fun checkName(name: String): Pair<Boolean, String?>
+}
+
+/**
+ * Checks if the name is CamelCase.
+ */
+open class RustCamelCaseNamingInspection(
+    elementType: String,
+    elementTitle: String = elementType
+) : RustNamingInspection(elementType, "a camel", elementTitle) {
+    override fun checkName(name: String): Pair<Boolean, String?> {
+        val str = name.trim('_')
+        if (!str.isEmpty() && str[0].canStartWord && '_' !in str) {
+            return Pair(true, null)
+        }
+        return Pair(false, if (str.isEmpty()) "CamelCase" else suggestName(name))
+    }
+
+    private fun suggestName(name: String): String {
+        val result = StringBuilder(name.length)
+        var wasUnderscore = true
+        var startWord = true
+        name.forEach { char ->
+            if (char == '_') {
+                wasUnderscore = true
+            } else if (wasUnderscore || startWord && char.canStartWord) {
+                result.append(char.toUpperCase())
+                wasUnderscore = false
+                startWord = false
+            } else {
+                startWord = char.isLowerCase()
+                result.append(char.toLowerCase())
+            }
+        }
+        return if (result.isEmpty()) "CamelCase" else result.toString()
+    }
+
+    private val Char.canStartWord: Boolean get() = isUpperCase() || isDigit()
+}
+
+/**
+ * Checks if the name is snake_case.
+ */
+open class RustSnakeCaseNamingInspection(elementType: String) : RustNamingInspection(elementType, "a snake") {
+    override fun checkName(name: String): Pair<Boolean, String?> {
+        val str = name.trim('_')
+        if (!str.isEmpty() && str.all { !it.isLetter() || it.isLowerCase() }) {
+            return Pair(true, null)
+        }
+        return Pair(false, if (str.isEmpty()) "snake_case" else name.toSnakeCase(false))
+    }
+}
+
+/**
+ * Checks if the name is UPPER_CASE.
+ */
+open class RustUpperCaseNamingInspection(elementType: String) : RustNamingInspection(elementType, "an upper") {
+    override fun checkName(name: String): Pair<Boolean, String?> {
+        val str = name.trim('_')
+        if (!str.isEmpty() && str.all { !it.isLetter() || it.isUpperCase() }) {
+            return Pair(true, null)
+        }
+        return Pair(false, if (str.isEmpty()) "UPPER_CASE" else name.toSnakeCase(true))
+    }
+}
+
+private fun String.toSnakeCase(upper: Boolean): String {
+    val result = StringBuilder(length + 3)     // 3 is a reasonable margin for growth when `_`s are added
+    result.append(takeWhile { it == '_' || it == '\'' })    // Preserve prefix
+    var firstPart = true
+    drop(result.length).splitToSequence('_').forEach pit@ { part ->
```

Comment:
Actually, I found it out when made a research after [your comment](https://github.com/intellij-rust/intellij-rust/pull/714#discussion_r82158230). So the thanks are mutual )